pkg/server/api_v2_databases_metadata.go

```diff
@@ -0,0 +1,385 @@
+	storeIdKey      = "storeId"
+	defaultPageSize = 10
+	defaultPageNum  = 1
+)
+
+// GetTableMetadata returns a paginated response of table metadata and statistics. This is not a live view of
+// the table data but instead is cached data that had been precomputed at an earlier time.
+//
+// The user making the request will receive table metadata based on the CONNECT database grant and admin privilege.
+// If the user provides a database id that they are not authorized for, they will receive an empty response. Similarly,
+// if the user does not provide a database id, the result set will only include databases in which they have the
+// previously mentioned authorizations.
+//
+// This endpoint supports the following query parameters:
+//
+// name: dbId
+// type: integer
+// description: The id of the database to fetch table metadata.
+//
+// name: name
+// type: string
+// description: a string which is used to match table and schema names against. This string is tokenized by "." and
+// all the tokens are matched against both schema and table name. For a table to match this parameter, each token must
+// match either the schema name or the table name.
+//
+// name: sortBy
+// type: string
+// description: Which column to sort by. This currently supports: "replicationSize", "ranges", "liveData", "columns",
+// "indexes", "lastUpdated". If a non supported value is provided, it will be ignored.
+//
+// name: sortOrder
+// type: string
+// description: The direction in which to order the sortBy column by. Supports either "asc" or "desc". If a non
+// supported value is provided, it will be ignored.
+//
+// name: pageSize
+// type: string
+// description: The size of the page of the result set to return.
+//
+// name: pageNum
+// type: string
+// description: The page number of the result set to return.
+//
+// name: storeId
+// type: integer
+// description: The id of the store to filter tables by. If the table contains data the store, it will be included in
+// the result set. Multiple storeId query parameters are support. If multiple are provided, a table will be included in
+// the result set if it contains data in at least one of the stores.
+func (a *apiV2Server) GetTableMetadata(w http.ResponseWriter, r *http.Request) {
+	ctx := r.Context()
+	ctx = a.sqlServer.AnnotateCtx(ctx)
+	sqlUser := authserver.UserFromHTTPAuthInfoContext(ctx)
+	// TODO (kyle): build http method handling directly into route registration
+	if r.Method != http.MethodGet {
+		http.Error(w, "Method Not Allowed", http.StatusMethodNotAllowed)
+		return
+	}
+
+	queryValues := r.URL.Query()
+	qs := struct {
+		dbId      apiutil.QueryStringVal[int]
+		tableName apiutil.QueryStringVal[string]
+		storeIds  apiutil.QueryStringVal[[]int]
+		sortBy    apiutil.QueryStringVal[string]
+		sortOrder apiutil.QueryStringVal[string]
+		pageNum   apiutil.QueryStringVal[int]
+		pageSize  apiutil.QueryStringVal[int]
+	}{
+		dbId:      apiutil.GetIntQueryStringVal(queryValues, dbIdKey),
```

Comment:
What are your thoughts about inlining the utility functions into a function for this struct?

```go
type tableMetadataQueryArgs  struct {
     dbId            int
     tableName string
     ...
}

func newTableMetadataQueryArgs(vals url.Values) (tableMetadataQueryArgs, error) {
    // if you want to check zero values
    dbIdStr := vals.Get(dbIdKey)
    dbId, err := strconv.Atoi(dbIdStr)
    if err != nil {
        return errors.Wrapf(err, "could not parse tableMetadata query arg %s", dbIdKey)
    }

     // if nil checks work better to verify existence
    var tableName *string
    if vals.Has(nameKey) {
        name := vals.Get(nameKey)
        tableName = &name
    }

    storeIds := make([]int, len(vals[storeIds])
    for _, storeIdStr := range vals[storeIds] {
         storeId, err := strconv.Atoi(storeIdStr)
         if err != nil {
             return errs.Wrapf(err, "could not parse tableMetadata storeId: %s", storeIdStr)
         }
         storeIds = append(storeIds)
    }
   

    return tableMetadataQueryArgs{
        tableName: tableName,
        ...
    }
} 
```

It's a bit wordy, but I think more clearly defines how the query will be parsed (absent an api specification)

Reply:
I see, yeah it makes sense since they're passed individually as arguments to the next function down - I think my only question is around the utility functions `GetIntQueryStringVal`. Do you have any thoughts about trying thing PR without them?

Reply:
Done

Reply:
Actually, I added GetIntQueryStringVal back in. I think not having this function results in a lot of duplicated code to get the param value and convert it to an int.